barretenberg/cpp/src/barretenberg/vm2/constraining/relations/emit_nullifier.test.cpp

```diff
@@ -0,0 +1,243 @@
+{
+    uint64_t prev_num_nullifiers_emitted = MAX_NULLIFIERS_PER_TX - 1;
+    TestTraceContainer trace({ {
+        { C::execution_sel_execute_emit_nullifier, 1 },
+        { C::execution_register_0_, /*nullifier=*/42 },
+        { C::execution_mem_tag_reg_0_, static_cast<uint8_t>(MemoryTag::FF) },
+        { C::execution_remaining_nullifiers_inv, FF(MAX_NULLIFIERS_PER_TX - prev_num_nullifiers_emitted).invert() },
+        { C::execution_sel_write_nullifier, 1 },
+        { C::execution_sel_opcode_error, 0 },
+        { C::execution_subtrace_operation_id, AVM_EXEC_OP_ID_EMIT_NULLIFIER },
+        { C::execution_prev_nullifier_tree_size, 1 },
+        { C::execution_nullifier_tree_size, 2 },
+        { C::execution_prev_num_nullifiers_emitted, prev_num_nullifiers_emitted },
+        { C::execution_num_nullifiers_emitted, prev_num_nullifiers_emitted + 1 },
+    } });
+    check_relation<emit_nullifier>(trace);
+}
+
+TEST(EmitNullifierConstrainingTest, LimitReached)
+{
+    uint64_t prev_num_nullifiers_emitted = MAX_NULLIFIERS_PER_TX;
+    TestTraceContainer trace({ {
+        { C::execution_sel_execute_emit_nullifier, 1 },
+        { C::execution_register_0_, /*nullifier=*/42 },
+        { C::execution_mem_tag_reg_0_, static_cast<uint8_t>(MemoryTag::FF) },
+        { C::execution_remaining_nullifiers_inv, 0 },
+        { C::execution_sel_write_nullifier, 0 },
+        { C::execution_sel_opcode_error, 1 },
+        { C::execution_subtrace_operation_id, AVM_EXEC_OP_ID_EMIT_NULLIFIER },
+        { C::execution_prev_nullifier_tree_size, 1 },
+        { C::execution_nullifier_tree_size, 1 },
+        { C::execution_prev_num_nullifiers_emitted, prev_num_nullifiers_emitted },
+        { C::execution_num_nullifiers_emitted, prev_num_nullifiers_emitted },
+    } });
+    check_relation<emit_nullifier>(trace);
+
+    // Negative test: sel_opcode_error must be on
+    trace.set(C::execution_sel_opcode_error, 0, 0);
+    EXPECT_THROW_WITH_MESSAGE(check_relation<emit_nullifier>(trace, emit_nullifier::SR_OPCODE_ERROR_IF_LIMIT_ERROR),
+                              "OPCODE_ERROR_IF_LIMIT_ERROR");
+
+    // Negative test: tree size must be the same
+    trace.set(C::execution_nullifier_tree_size, 0, 2);
+    EXPECT_THROW_WITH_MESSAGE(
+        check_relation<emit_nullifier>(trace, emit_nullifier::SR_EMIT_NULLIFIER_TREE_SIZE_INCREASE),
+        "EMIT_NULLIFIER_TREE_SIZE_INCREASE");
+
+    // Negative test: num nullifiers emitted must be the same
+    trace.set(C::execution_num_nullifiers_emitted, 0, prev_num_nullifiers_emitted + 1);
+    EXPECT_THROW_WITH_MESSAGE(
+        check_relation<emit_nullifier>(trace, emit_nullifier::SR_EMIT_NULLIFIER_NUM_NULLIFIERS_EMITTED_INCREASE),
+        "EMIT_NULLIFIER_NUM_NULLIFIERS_EMITTED_INCREASE");
+}
+
+TEST(EmitNullifierConstrainingTest, Interactions)
+{
+    NiceMock<MockPoseidon2> poseidon2;
+    NiceMock<MockMerkleCheck> merkle_check;
+    NiceMock<MockRangeCheck> range_check;
+    NiceMock<MockFieldGreaterThan> field_gt;
```

Comment:
We should migrate these to `StrictMock`s. It looks like we actually do care about what these things do (looking at the `EXPECT`s that follow). This will cover us in case there are any missed expectations whose default values might impact our control flow.

FYI there are also now "fakes" (`FakePoseidon` and `FakeGt`) that might be helpful for you too.

Reply:
I said this in a standup when you were OOO: NiceMock is almost always not what you want. It returns some default value which can change your control flow. StrictMock is what you want. If you are making 0 calls, strictmock will be fine. If you are making > 0 calls, then it is expected that the result of the call modifies your control flow so you want to set an EXPECT_CALL. The only case that might make sense for a NiceMock is when you call some `void assert_...` which will not change anything, but ideally you can just use StrickMock unless it becomes too verbose.